### PR TITLE
HDDS-11117. Introduce debug CLI command to show the value schema of any rocksDB

### PR DESCRIPTION
**Pull request by Tejaskriya**
## What changes were proposed in this pull request?

RocksDB stores data in key-value pairs. The value itself may have some kind of key-value structure. A command is needed which shows what structure this value has and prints the same. This would help in parsing and filtering the output of `ozone debug ldb scan` command to obtain only the required details from the value.
In this PR, A command `ozone debug ldb --db=<db-path> value-schema --cf=<table-name>` has been introduced, which uses java reflections to fetch the fields of the value type of a rocksDB for all levels. `--depth` option can be used to limit the hierarchy till which it fetches the fields.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-11117

## How was this patch tested?

Tested manually in docker setup, also added test in TestLDBCli. Sample output:
```
sh-4.2$ ozone debug ldb --db=/data/metadata/om.db value-schema --cf=keyTable --depth=1
{
  "OmKeyInfo" : {
    "bucketName" : "String",
    "metadata" : "struct",
    "fileName" : "String",
    "creationTime" : "long",
    "isFile" : "boolean",
    "acls" : "struct",
    "keyName" : "String",
    "replicationConfig" : "struct",
    "encInfo" : "struct",
    "dataSize" : "long",
    "tags" : "struct",
    "keyLocationVersions" : "struct",
    "updateID" : "long",
    "ownerName" : "String",
    "modificationTime" : "long",
    "parentObjectID" : "long",
    "volumeName" : "String",
    "fileChecksum" : "struct",
    "objectID" : "long"
  }
}

sh-4.2$ ozone debug ldb --db=/data/metadata/om.db value-schema --cf=keyTable 
{
  "OmKeyInfo" : {
    "bucketName" : "String",
    "metadata" : { },
    "fileName" : "String",
    "creationTime" : "long",
    "isFile" : "boolean",
    "acls" : {
      "toStringMethod" : { },
      "hashCodeMethod" : { },
      "name" : "String",
      "type" : {
        "name" : "String",
        "value" : "String",
        "ordinal" : "int"
      },
      "aclScope" : {
        "name" : "String",
        "ordinal" : "int"
      },
      "aclBits" : "int"
    },
    "keyName" : "String",
    "replicationConfig" : { },
    "encInfo" : {
      "ezKeyVersionName" : "String",
      "keyName" : "String",
      "edek" : { },
      "cipherSuite" : {
        "unknownValue" : {
          "value" : "int"
        },
        "name" : "String",
        "algoBlockSize" : "int",
        "ordinal" : "int"
      },
      "version" : {
        "unknownValue" : {
          "value" : "int"
        },
        "name" : "String",
        "description" : "String",
        "version" : "int",
        "ordinal" : "int"
      },
      "iv" : { }
    },
    "dataSize" : "long",
    "tags" : { },
    "keyLocationVersions" : {
      "isMultipartKey" : "boolean",
      "locationVersionMap" : { },
      "version" : "long"
    },
    "updateID" : "long",
    "ownerName" : "String",
    "modificationTime" : "long",
    "parentObjectID" : "long",
    "volumeName" : "String",
    "fileChecksum" : { },
    "objectID" : "long"
  }
}
```

### CODE REVIEW

**Comment by sadanand48** [comment]
The current approach looks good and I'm okay if we go with it, but since we use proto for encoding and decoding , we could also leverage protobuf's [getDescriptor](https://protobuf.dev/reference/java/api-docs/com/google/protobuf/Descriptors) to print the schema . Something like this would print 1 level of schema for the message, All we have to do is find the proto object for each valueType. We probably would need to define toProtoBuf() method for each message type to convert to proto which would be extra work.  Please check if this is easier or else the current patch looks good.
```java
    OzoneManagerProtocolProtos.KeyInfo keyInfo = OzoneManagerProtocolProtos.KeyInfo.newBuilder()
        .getDefaultInstanceForType();
    System.out.println(keyInfo.getDescriptorForType().toProto());
```

**Comment by Tejaskriya** [comment]
> The current approach looks good and I'm okay if we go with it, but since we use proto for encoding and decoding , we could also leverage protobuf's [getDescriptor](https://protobuf.dev/reference/java/api-docs/com/google/protobuf/Descriptors) to print the schema . Something like this would print 1 level of schema for the message, All we have to do is find the proto object for each valueType. We probably would need to define toProtoBuf() method for each message type to convert to proto which would be extra work. Please check if this is easier or else the current patch looks good.

With this approach, one major issue was that the field names were different from what is seen when we do a `ozone debug ldb scan` command on a table. That is due to the variable name difference in proto message definition and the java class. For example, in the volumeTable, in `scan` we see it as "acls" but the proto message is "volumeacls".
The current approach seems to be simpler IMO.

**Comment by Tejaskriya** [comment]
@sumitagrawl @sadanand48 I have updated the patch to include a --depth option using which we can specify in what depth we want the schema (older patch just showed 2 levels). Also, the command name is changed to `value-schema` so that it is not confused as some table schema type of operation.
Could you please review the updated patch?

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/debug/ValueSchema.java [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+    }
+    final DBColumnFamilyDefinition<?, ?> columnFamilyDefinition =
+        dbDefinition.getColumnFamily(tableName);
+    if (columnFamilyDefinition == null) {
+      err().print("Error: Table with name '" + tableName + "' not found");
+      return false;
+    }
+
+    Class<?> c = columnFamilyDefinition.getValueType();
+    valueSchema.put(c.getSimpleName(), getFieldsStructure(c, depth));
```
Comment (sumitagrawl) [inline hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/debug/ValueSchema.java:117]: Do avoid any recursion issue, maxDepth can be updated to "10", IMO, no more than this depth filter is required

**Review by sumitagrawl — COMMENTED** [review]
@Tejaskriya overall LGTM, have one minor comment

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/debug/ValueSchema.java (hunk 1 above)
Comment (Tejaskriya) [reply hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/debug/ValueSchema.java:121]: Thank you for the review! I have made this change as you suggested now. Could you please review it again?

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/debug/ValueSchema.java [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+
+    return true;
+  }
+
+  private Object getFieldsStructure(Class<?> clazz, int currentDepth) {
+    if (clazz.isPrimitive() || String.class.equals(clazz)) {
+      return clazz.getSimpleName();
+    } else if (currentDepth == 0) {
```
Comment (sumitagrawl) [inline hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/debug/ValueSchema.java:129]: I think we should allow depth "0" value as current logic for first level

**Review by sumitagrawl — APPROVED** [review]
LGTM

**Review by sadanand48 — APPROVED** [review]
Thanks @Tejaskriya for the patch, LGTM too.

**Comment by Tejaskriya** [comment]
Thanks @sadanand48 and @sumitagrawl for the reviews and approvals! 
All the checks have come up green.

**Pull request merged**